### ISSUE TITLE
eliminate memcpy during encryption for short packets

### ISSUE THREAD

**Issue by cpacejo** [org-zpr/zpr-core#385]
"Short" packets (i.e. practically all packets) can simply be encrypted into a different location in the source buffer.

**Comment by ZprBot1** [comment]
## Plan

### What the code does today

`encrypt_full()` in `adapter/ph/src/fastpath.rs` (called on every non-transit egress packet) does:

```rust
let mut enc_buf = [0u8; config::PACKET_BUFFER_SIZE];   // 12288 bytes, zeroed
codec.encrypt_transport_stateless(&pkt.body()[1..], &mut enc_buf)?;
pkt.shrink_by(encr_len);       // drop cleartext body, keep ZPI byte
pkt.put(&enc_buf[0..len]);     // copy ciphertext back into the packet
```

So each packet pays **two** avoidable costs:

1. a 12 KiB stack zeroing (`[0u8; PACKET_BUFFER_SIZE]`, `PACKET_BUFFER_SIZE = 4096*3`) — this is
   larger than the packet itself for essentially all traffic, and it is not optimised away because
   the buffer's address escapes into a `dyn Codec` call;
2. a full `memcpy` of the ciphertext from the scratch buffer back into the packet buffer.

`km::encrypt_transport_zdp()` and `decrypt_full()` have the same shape.

### Proposed change

Encrypt directly into the packet's own backing buffer, at a different offset, exactly as the issue
describes. A `Packet` buffer is `PACKET_BUFFER_SIZE` bytes with the body at `metadata.offset`, so for
any packet short enough there is unused **tailroom** after the body that cannot overlap the
plaintext:

```
before:  [ metadata ][ headroom ][ ZPI ][ plaintext .......... ][ tailroom ................ ]
after:   [ metadata ][ ........ dead ......................... ][ZPI][ ciphertext ][ tail   ]
```

Steps in `encrypt_full()`:

1. Compute required space = `encr_len + codec.max_overhead()` and compare against the packet's
   current tailroom (`BufMut::remaining_mut()`).
2. **Fast path** (fits — i.e. "short" packets, practically all of them): split the backing buffer
   into `(&plaintext, &mut tailroom)` and hand those two non-overlapping slices to
   `encrypt_transport_stateless()`. Then move the 1-byte ZPI header to sit immediately before the
   ciphertext and set `offset`/`len` accordingly. Net data movement: **1 byte**, no zeroing.
3. **Slow path** (does not fit): keep today's scratch-buffer behaviour so oversized packets are
   still handled identically. This path stays correct but is no longer on the common path.

Supporting changes:

* Add `fn max_overhead(&self) -> usize` to the `Codec` trait — the number of bytes ciphertext may
  exceed plaintext by. `NoiseCodec` returns `NOISE_PADLEN` (8-byte nonce + 16-byte tag = 24),
  `NullCodec`/`UnimplCodec` return 0. This is needed because the trait currently gives the caller no
  way to size an output buffer; `decrypt_full()` already works around that by hardcoding
  `NOISE_PADLEN` at the call site, which is a latent bug the moment a second codec exists.
* Add a `Packet` accessor that yields the body and the tailroom as two disjoint mutable/immutable
  slices (a thin `split_at_mut` wrapper), since `body()` and `chunk_mut()` cannot be held at once.

### Deliberately out of scope

* `decrypt_full()` and `km::encrypt_transport_zdp()`/`decrypt_transport_zdp()` carry the same 12 KiB
  scratch buffer. The issue is scoped to encryption; I would rather land this shape, get it reviewed,
  and then apply the same treatment to the decrypt side in a follow-up than mix them in one PR.
  Say the word if you would rather have all of them in this PR.

### Verification

* Round-trip unit tests in `fastpath.rs` with a codec that actually expands (nonce + tag), asserting
  the ZPI byte survives, the ciphertext is byte-identical to the scratch-buffer implementation, and
  `encrypt_full` -> `decrypt_full` returns the original plaintext.
* An explicit test that a packet with insufficient tailroom takes the fallback path and still
  produces the same bytes.
* Full gate: `cargo build --all-targets`, `cargo fmt --check`, `cargo test`, `-D warnings`.

### One question

Is there anything downstream of egress encryption that relies on the packet retaining tailroom or
its original `offset`? My reading of `substrate_egress()` is no — the packet goes straight onto
`substrate_egress_q` and only `body()` is written to the socket — but the fast path does consume the
tailroom, so I want that confirmed rather than assumed. I am proceeding on the assumption that it is
safe.

**Comment by ZprBot1** [comment]
Implementation is complete and the full CI gate passes locally (`cargo build --all-targets`, `cargo fmt --check`, `cargo test` — 222 passed, `-D warnings` clean), but I cannot push the branch:

```
ERROR: Permission to org-zpr/zpr-core.git denied to ZprBot1.
```

`gh api repos/org-zpr/zpr-core` confirms it is a repo permission, not an auth or SSH problem:

| repo | pull | push | triage |
|---|---|---|---|
| zpr-core | true | **false** | false |
| zpr-compiler | true | true | true |

SSH authenticates fine (`Hi ZprBot1!`) and I am a member of `core-devs`, `designers`, `robots` and `zpr-contributors` — so whichever team carries write on zpr-compiler does not carry it on zpr-core. Granting ZprBot1 (or the appropriate team) **write** on org-zpr/zpr-core unblocks it; the commit is ready locally and I will push and open the PR immediately.

Measured effect of the change, for context, on a 64-byte body with a codec that expands by 24 bytes (release build, buffer allocation excluded from both paths):

```
in-place        28 ns/pkt
scratch buffer 192 ns/pkt
```

**Comment by cpacejo** [comment]
Correct, nothing downstream of `substrate_egress()` relies on there being remaining tailroom, or on the packet remaining in the same position.  The same is true in the ingress direction.

Regarding decryption, yes, doing that in a separate PR is fine.
